zebra-network/src/address_book.rs

```diff
@@ -463,6 +514,13 @@ impl AddressBook {
         self.by_addr.descending_values().cloned()
     }
 
+    /// Returns true if there is a recently live peer connection with the provided IP address.
+    fn has_active_peer_with_ip(&self, ip: &IpAddr, chrono_now: chrono::DateTime<Utc>) -> bool {
+        self.most_recent_by_ip
+            .get(ip)
+            .map_or(false, |peer| peer.was_recently_live(chrono_now))
```

Comment:
Updated to check last_attempt, last_response, and last_failure times in https://github.com/ZcashFoundation/zebra/pull/7029/commits/489e63d14a9c9db1796c275c60234f832b7f497a.

Reply:
Why does last_failure matter?

Reply:
So that `reconnection_peers()` skips addresses with IPs where there's a recently failed connection as it would if the connection itself had recently failed (in `is_ready_for_connection_attempt()`).

Reply:
I think that checking the failure time enables an attack which pollutes our address book with fake peers on the same IP as a real peer. If the real peer is on a non-default port, the fake peers will all be tried before it, delaying a connection to the real peer by 2 minutes per fake peer. (Which is effectively never, because each IP address has 65000 TCP ports, so the limit is the 4000 address book addresses.)

So I think we should treat recently successful connections differently to connections that were never successful and immediately failed.

Specifically, an outbound connection to the same IP should not be attempted when:
* the peer recently had a successful connection and responded (~ 2 minute timeout)

But an outbound connection to the same IP should be attempted even if:
* we recently attempted a handshake to that peer
* we recently failed a connection to that peer without completing the handshake (`last_response` was not updated)

If we want to do anything about avoiding attempting multiple connections to the same IP at the same time, let's do that in a separate ticket with a lower priority?

Reply:
> But an outbound connection to the same IP should be attempted even if:
> 
> - we recently attempted a handshake to that peer
> - we recently failed a connection to that peer without completing the handshake (last_response was not updated)

Fixed in https://github.com/ZcashFoundation/zebra/pull/7029/commits/788bce3b9ea89d3d5b3d535d09720b9b04c169a0 and https://github.com/ZcashFoundation/zebra/pull/7029/commits/f8338f1871f829851582fd3d7015effa7f0e04be.